nitro-src/claim.ts

```diff
@@ -1,163 +1,219 @@
-    assetIndex++
-  ) {
-    const guarantees = initialGuaranteeOutcome[assetIndex].allocations;
-    const targetAllocations = initialTargetOutcome[assetIndex].allocations;
+  // Deep-copy as we will mutating these objects.
+  const afterClaimGuarantee: Exit = _.cloneDeep(guarantees);
+  const afterClaimHoldings = _.cloneDeep(holdings);
+
+  const afterClaimTargetOutcome: Exit = [];
+  const afterClaimExits: Exit = [];
 
-    const updatedAllocations = targetAllocations.map((a) =>
-      JSON.parse(JSON.stringify(a))
-    ); // copy the allocations for mutation
+  // Iterate through every asset
+  for (let assetIndex = 0; assetIndex < guarantees.length; assetIndex++) {
+    const guaranteesForOneAsset = guarantees[assetIndex].allocations;
+    const targetAllocations = targetOutcome[assetIndex].allocations;
 
     const singleAssetExit: SingleAssetExit = {
-      ...initialTargetOutcome[assetIndex],
+      ...targetOutcome[assetIndex],
       allocations: [], // start with an empty array
     };
 
     if (
-      guarantees[targetChannelIndex].allocationType !== AllocationType.guarantee
-    )
-      throw Error;
+      guaranteesForOneAsset[targetChannelIndex].allocationType !==
+      AllocationType.guarantee
+    ) {
+      throw Error("Expected allocation type guarantee");
+    }
 
-    let surplus = BigNumber.from(initialHoldings[assetIndex]);
     // Any guarantees before this one have priority on the funds
     // So we must account for that by reducing the surplus
-    for (
-      let guaranteeIndex = 0;
-      guaranteeIndex < targetChannelIndex;
-      guaranteeIndex++
-    ) {
-      const { amount } = guarantees[guaranteeIndex];
+    const guaranteesBeforeOurs = guaranteesForOneAsset.slice(
+      0,
+      targetChannelIndex
+    );
+    const amountBeforeUs = sumAllocationAmounts(guaranteesBeforeOurs);
+    let maxAmountCanPayOut = holdings[assetIndex].sub(amountBeforeUs);
 
-      surplus = surplus.sub(min(BigNumber.from(amount), surplus)); // Prevent going below 0
-    }
-    // If there are not enough funds to fund the guarantee we return immediately
-    if (surplus.isZero()) {
-      return {
-        updatedGuaranteeOutcome,
-        updatedHoldings,
-        updatedTargetOutcome,
-        exit,
-      };
+    // If there are not enough funds to fund the guarantee, we skip the asset
+    if (maxAmountCanPayOut.lte(0)) {
+      continue;
     }
 
-    surplus = min(
-      surplus,
-      BigNumber.from(guarantees[targetChannelIndex].amount)
+    maxAmountCanPayOut = min(
+      maxAmountCanPayOut,
+      BigNumber.from(guaranteesForOneAsset[targetChannelIndex].amount)
     );
 
-    let exitRequestIndex = 0;
+    const {
+      afterClaimAllocations,
+      afterClaimGuaranteeAmount,
+      newExits,
+    } = claimOneGuaranteeForOneAsset(
+      guaranteesForOneAsset[targetChannelIndex],
+      targetAllocations,
+      maxAmountCanPayOut,
+      exitRequest[assetIndex]
+    );
```

Comment:
The deleted line is preventing a suggestion here, but maybe:
```
const amountToPayOut = min(
  maxAmountCanPayOut,
  BigNumber.from(guaranteesForOneAsset[targetChannelIndex].amount)
);

const {
  afterClaimAllocations,
  afterClaimGuaranteeAmount,
  newExits,
} = claimOneGuaranteeForOneAsset(
  guaranteesForOneAsset[targetChannelIndex],
  targetAllocations,
  amoutToPayOut,
  exitRequest[assetIndex]
);
```

Reply:
Reading further, is this `min` check necessary here? `claimOneGuaranteeForOneAsset` manages underfunded claims itself, and its signature asks for `holdingsForGuarantee`, which I interpret as being the same as `maxAmountCanPayOut` here.

Reply:
I think it might be necessary. In the `for` loop in the `claim` function, we are calculating the maximum amount of funds that are available for a guarantee. In the `claimOneGuaranteeForOneAsset`, we are distributing this maximum amount between all of the destinations in the guarantee. There is perhaps a better way to express these calculations!